nomad/core_sched.go

```diff
@@ -28,29 +28,42 @@ type CoreScheduler struct {
 	snap   *state.StateSnapshot
 	logger log.Logger
 
-	// custom GC Threshold values can be used by unit tests to simulate time
-	// manipulation
-	customJobGCThreshold                time.Duration
-	customEvalGCThreshold               time.Duration
-	customBatchEvalGCThreshold          time.Duration
-	customNodeGCThreshold               time.Duration
-	customDeploymentGCThreshold         time.Duration
-	customCSIVolumeClaimGCThreshold     time.Duration
-	customCSIPluginGCThreshold          time.Duration
-	customACLTokenExpirationGCThreshold time.Duration
-	customRootKeyGCThreshold            time.Duration
+	// custom GC Threshold is a map of object names to threshold values which can be
+	// used by unit tests to simulate time manipulation, and is used by forceGC to
+	// set minimal threshold so that virtually all eligible objects will get GCd
+	customGCThreshold map[string]time.Duration
 }
 
 // NewCoreScheduler is used to return a new system scheduler instance
 func NewCoreScheduler(srv *Server, snap *state.StateSnapshot) scheduler.Scheduler {
 	s := &CoreScheduler{
-		srv:    srv,
-		snap:   snap,
-		logger: srv.logger.ResetNamed("core.sched"),
+		srv:               srv,
+		snap:              snap,
+		logger:            srv.logger.ResetNamed("core.sched"),
+		customGCThreshold: make(map[string]time.Duration),
 	}
 	return s
 }
 
+func (c *CoreScheduler) setCustomThresholdForObject(objectName string, threshold time.Duration) {
+	c.customGCThreshold[objectName] = threshold
+}
+
+func (c *CoreScheduler) setCustomThresholdForAllObjects(threshold time.Duration) {
+	for _, objectName := range []string{
+		"job",
+		"eval",
+		"batchEval",
+		"deployment",
+		"csiPlugin",
+		"csiVolume",
+		"token",
+		"node",
+	} {
+		c.setCustomThresholdForObject(objectName, threshold)
+	}
+}
```

Comment:
My main concern with this kind of approach is that we're relying on matching strings to tables/objects to funcs.

What if instead we passed the threshold into the funcs directly from Process? Process already does string parsing of the Eval (which is a different string than this map keys off of!), so I don't think it would make Process too much more complicated to pass the appropriate threshold from config into funcs. It _might_ make testing easier to? You can pass thresholds in directly to gc funcs or you can customize the config. I would hope that would cover every case.

Reply:
> Since we don't add new gc-able objects often, it seems even more likely we'd forget the force gc path.

That is, sadly, very likely. 

> What if instead we passed the threshold into the funcs directly from Process?

I looked at that but it's a major change. Not only would we have to change every call ro `Process` (of which there are _many_), but we'd also need to change the scheduler interface. Not saying we shouldn't do this, just saying it's a big change.

Reply:
> Not only would we have to change every call to Process

Do we have to call Process or just change the code _inside_ [CoreScheduler.Process](https://github.com/hashicorp/nomad/blob/13657e02dae2e536d2b510e62efdd26dc2d8f3df/nomad/core_sched.go#L58-L81)?

For example could the call to `c.jobGC(eval)` be changed to `c.jobGC(eval, c.srv.config.JobGCThreshold)` to avoid the config lookup inside `jobGC` itself?

Reply:
The passed in threshold could still be overridden by custom values, so I don't think it would break tests.  Although it does feel weird to pass in a struct field thats already available to the receiver.
Which makes me wonder if (I think your original idea) of just passing in a forceGC boolean might be the best way to go here.

Reply:
Right. Passing a `force` boolean to every GC method will still be very messy. Because I can't think of a better way of forcing than setting a very very small threshold value. So then for every GC method we'd have to check 3 places where a threshold might've been set: config, custom field in the scheduler, or force bool induced minimal value. That, to me, is even less elegant.

Reply:
@schmichael @mismithhisler Gentlemen I heard your complaints, and I present to you 1d149a0. 

We no longer keep stringy hash maps, and we simplify the way force GC happens. And yet we still let people test "time manipulation" by forcing custom threshold values. See if it pleases you.

Reply:
I am pleased.